felix/bpf/proxy/syncer.go

```diff
@@ -552,7 +552,7 @@ func (s *Syncer) apply(state DPSyncerState) error {
 		eps := make([]k8sp.Endpoint, 0, len(state.EpsMap[sname]))
 		for _, ep := range state.EpsMap[sname] {
 			zoneHints := ep.GetZoneHints()
-			if ep.IsReady() {
+			if ep.IsReady() || ep.IsTerminating() {
```

Comment:
Does applySvc filter out Terminating endpoints later?  Terminating endpoints shouldn't be in the BPF maps.

Reply:
Also, placing `IsTerminating()` next to `IsReady()` implies that if endpoint became not ready, in-flight requests would also be dropped? Or I am just being crazy here?

Reply:
Thanks @fasaxc - [applySvc](https://github.com/projectcalico/calico/blob/master/felix/bpf/proxy/syncer.go#L349)() does not seem to filter out Terminating endpoints.  My understanding of the current syncer code in master is that all code paths seem to be channelled thru the [apply](https://github.com/projectcalico/calico/blob/master/felix/bpf/proxy/syncer.go#L526)() function which initially filters out only endpoints at status [Ready](https://github.com/projectcalico/calico/blob/master/felix/bpf/proxy/syncer.go#L555) and adds them to the eps collection for further processing.  

I have tried to grok other parts of the code and it seems that if an endpoint is at status other than Ready during the apply() function then it would not be included here thus not have an opportunity to be processed later on in the code - hence the modification in the PR but pls correct me if I am wrong about this - thx.

Reply:
I think, checking the terminating flag is the simplest (nice!) way to tackle this problem. However,this will need to be a little bit more complicated.

(1) yes, we need to let the terminating through here so that we know later on that they are terminating.
(2) we must remove the terminating from the bpf maps because we should not allow new connections to that endpoint via the service as `At this point, the pod stops getting new traffic. Containers running in the pod will not be affected.`
(3) we need to remember that the pod is terminating to answer `true` when [this callback](https://github.com/projectcalico/calico/blob/master/felix/bpf/proxy/syncer.go#L1153) is called.

Perhaps we need to keep the state in [this struct](https://github.com/projectcalico/calico/blob/master/felix/bpf/proxy/syncer.go#L43) so that when we (a) are writing out the bpf backend map we skip the terminating and (b) when answering the question whether that backend exists, we know about it.

Reply:
Update code to prevent terminating pods from being added to the bpf maps as outlined above @tomastigera

Reply:
Implemented changes as per thread